solr/core/src/test/org/apache/solr/search/TestCpuAllowedLimit.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.solr.SolrTestCaseJ4;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.util.SolrJettyTestRule;
+import org.apache.solr.util.ThreadCpuTimer;
+import org.junit.AfterClass;
+import org.junit.Assume;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TestCpuAllowedLimit extends SolrTestCaseJ4 {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  @ClassRule public static final SolrJettyTestRule solrRule = new SolrJettyTestRule();
+
+  @BeforeClass
+  public static void setup() {
+    System.setProperty(ThreadCpuTimer.ENABLE_CPU_TIME, "true");
+  }
+
+  @AfterClass
+  public static void teardown() {
+    System.clearProperty(ThreadCpuTimer.ENABLE_CPU_TIME);
+  }
+
+  @Test
+  public void testCompareToWallClock() throws Exception {
+    Assume.assumeTrue("Thread CPU time monitoring is not available", ThreadCpuTimer.isSupported());
+    long limitMs = 100;
+    CpuAllowedLimit cpuLimit = new CpuAllowedLimit(limitMs);
+    int[] randoms = new int[100];
+    long startNs = System.nanoTime();
+    int wakeups = 0;
+    while (!cpuLimit.shouldExit()) {
+      Thread.sleep(1);
+      // do some busywork
+      for (int i = 0; i < randoms.length; i++) {
+        randoms[i] = random().nextInt();
+      }
+      wakeups++;
+    }
+    long endNs = System.nanoTime();
+    long wallTimeDeltaMs = TimeUnit.MILLISECONDS.convert(endNs - startNs, TimeUnit.NANOSECONDS);
+    log.info(
+        "CPU limit: {} ms, elapsed wall-clock: {} ms, wakeups: {}",
+        limitMs,
+        wallTimeDeltaMs,
+        wakeups);
+    assertTrue(
+        "Elapsed wall-clock time expected much larger than 100ms but was " + wallTimeDeltaMs,
+        limitMs < wallTimeDeltaMs);
+  }
+
+  private Path createConfigSet() throws Exception {
+    Path configSet = createTempDir();
+    copyMinConf(configSet.toFile());
+    // insert an expensive search component
+    Path solrConfig = configSet.resolve("conf/solrconfig.xml");
+    Files.writeString(
+        solrConfig,
+        Files.readString(solrConfig)
+            .replace(
+                "<requestHandler",
+                "<searchComponent name=\"expensiveSearchComponent\"\n"
+                    + "                   class=\"org.apache.solr.search.ExpensiveSearchComponent\"/>\n"
+                    + "\n"
+                    + "  <requestHandler")
+            .replace(
+                "class=\"solr.SearchHandler\">",
+                "class=\"solr.SearchHandler\">\n"
+                    + "    <arr name=\"first-components\">\n"
+                    + "      <str>expensiveSearchComponent</str>\n"
+                    + "    </arr>\n"));
+    return configSet.resolve("conf");
+  }
+
+  @Test
+  public void testDistribLimit() throws Exception {
```

Comment:
no distributed-search  (e.g. via "shards")?  Not mandatory to test this but if I thought that's why you wanted SolrCloudTestCase.

Reply:
Hmm... yes, that would be useful. `TestCpuTimeSearch` has an example, I'm going to use it (though it's concerning that the distrib test there is Ignored).

Reply:
Reverted back to SolrCloudTestCase, the other method was too unstable.